### ISSUE TITLE
issue with ATMEGA480X or DX CPU and libraries

### ISSUE THREAD

**Issue by LaurentR59** [aikopras/AP_DCC_library#1]
Hello
I tried to use libraire and exemple code with MegacoreX CPU like ATMEGA4809 or DX (128DB32)
I get some issue on compiling due to some issue identified from your lib (mostly suspected)
Your code is maching with:
Event& myEvent = Event::assign_generator_pin(dccPin);
but on librarie file def is a bit different like:
Event& Event::assign_generator_pin(uint8_t port, uint8_t port_pin) {
So need some adjustment I think or used function is not the good one at this step

A bit later on compile issue i get this
:...\Arduino\libraries\AP_DCC_Library\src\sup_isr_MegaCoreX_DxCore.h:193:28: error: 'tcb0_capt' is not a member of 'user'
     myEvent.set_user(user::tcb0_capt);
but probably linked with initial issue observed.

May it coud be fixed soon?

Please not that compiling with hardware as NANo_EVRY work like a charm including 328 mode or 480x mode. but unable to do it for current CPU like 480X natively.

Regards
Laurent

**Comment by LaurentR59** [comment]
Hi
As per illustration note i used exemple Loco_serial_Monitor and get this resuklt with MEGACOREX 1.0.9 librarie

In file included from C:\Users\laure\Documents\Arduino\libraries\AP_DCC_Library\src\sup_isr.cpp:44:0:
C:\Users\laure\Documents\Arduino\libraries\AP_DCC_Library\src\sup_isr_MegaCoreX_DxCore.h: In member function 'void DccMessage::initEventSystem(uint8_t)':
C:\Users\laure\Documents\Arduino\libraries\AP_DCC_Library\src\sup_isr_MegaCoreX_DxCore.h:182:27: error: 'assign_generator_pin' is not a member of 'Event'
   Event& myEvent = Event::assign_generator_pin(dccPin);
                           ^~~~~~~~~~~~~~~~~~~~
C:\Users\laure\Documents\Arduino\libraries\AP_DCC_Library\src\sup_isr_MegaCoreX_DxCore.h:193:28: error: 'tcb0_capt' is not a member of 'user'
     myEvent.set_user(user::tcb0_capt);
                            ^~~~~~~~~
C:\Users\laure\Documents\Arduino\libraries\AP_DCC_Library\src\sup_isr_MegaCoreX_DxCore.h:193:28: note: suggested alternative:
In file included from C:\Users\laure\Documents\Arduino\libraries\AP_DCC_Library\src\sup_isr_MegaCoreX_DxCore.h:54:0,
                 from C:\Users\laure\Documents\Arduino\libraries\AP_DCC_Library\src\sup_isr.cpp:44:
C:\Users\laure\AppData\Local\Arduino15\packages\MegaCoreX\hardware\megaavr\1.0.9\libraries\Event\src/Event.h:35:5: note:   'tcb0_capt'
     tcb0_capt     = 0xA0,

Hope it could help a bit more...

Laurent

**Comment by aikopras** [comment]
Hi Laurent

Thanks.If there are errors in my library, I will certainly fix them.
I guess, however, that the problem is caused by the versions you have for the Event library (both on DxCore and MegaCoreX). 

Let's take as example megaCoreX. You run release V1.0.9, which was released on August 12, 2021. However,  the new Event Library was committed to GitHub by MCUdude on August 21, so after the release of V1.0.9. 
If you install MegaCoreX / DxCore via the Arduino Boards library, you get the latest official release, but WITHOUT the commits that got added later. Therefore I guess you still run the old version of the Event Library.

To get the latest Event Library (which is needed for my code), on your computer you need to replace the "release version" with the version you can directly download from the GitHub web repository:
[https://github.com/MCUdude/MegaCoreX/tree/master/megaavr/libraries](https://github.com/MCUdude/MegaCoreX/tree/master/megaavr/libraries) 

Let me know if this solved the problem, or if you need help to make this replacement.
Alternatively you can ask the authors of MegaCoreX and DxCore to issue a new release (that will automatically include all commits since the last release).

**Comment by LaurentR59** [comment]
Hi Aiko
Thanks for your message.
Anyway it steel KO after some adjustment.
I observed on MEGACOREX git that latest version for EVENT is more recent than yours proposal ( 2months ago) so may it could introduce some minor change especially on Event& Event::assign_generator_pin(uint8_t port, uint8_t port_pin) { whitch is the current version.

Please put a look on file content to check it. @ https://github.com/MCUdude/MegaCoreX/blob/master/megaavr/libraries/Event/src/Event.cpp

Laurent

**Comment by aikopras** [comment]
Hi Laurent
I've been using the Event Lib that MCUdude (and SpenceKonde) developed outside the MegaCoreX main branch. That allowed me to test and modify my code while the Event Lib was being developed. After my code was released, MCUdude merged the EventLib into the MegaCoreX main branch. That explains the time difference between his release and my release. 
See also the discussion on: [https://github.com/MCUdude/MegaCoreX/issues/126](https://github.com/MCUdude/MegaCoreX/issues/126) 

Bye, Aiko

**Comment by LaurentR59** [comment]
Hello Aiko
Thanks for the provided link. It s pretty hard to follow all steps due to my newbie experience anyway!

Although
I performed a full fresh install of ARDUINO IDE 1.8.19
I mannually get zip from your LIB and MEGACOREX, DXCORE,mithycore
unzipped and place them into ...document/arduino/libraries
 updated cards management up to latest version for each lib provided earlier

open your exemple and select some CPU (4809 and 4808 with some spec)  to compile and... miracle! it was fixed and go up to end.

Now i need to run same behavior on PlateformIO....

sic! hope to succeed in such operation!

Laurent

**Comment by aikopras** [comment]
Hi Laurent

Glad to hear this worked. It would be quite convenient if MCUdude would create a new release, since then things will work out of the box. Maybe you can ask him (I guess it would make sense to do that); probably platformIO would then become easier as well (I have no experience there, however).
In the beginning it was for me hard and confusing as well, but after a relative short time it became clear and I could find these issues myself as well. Some welcome to the club :-)

Bye, Aiko

**Comment by aikopras** [comment]
Since this issue is resolved, I'll close it.
